src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.DateAndTime.LanguageServices
+{
+    /// <summary>
+    /// Helper class to detect <see cref="System.DateTime"/> and <see cref="DateTimeOffset"/> format
+    /// strings in a document efficiently.
+    /// </summary>
+    internal sealed class DateAndTimePatternDetector
+    {
+        private const string FormatName = "format";
+
+        private static readonly ConditionalWeakTable<SemanticModel, DateAndTimePatternDetector?> _modelToDetector =
+            new ConditionalWeakTable<SemanticModel, DateAndTimePatternDetector?>();
+
+        private readonly EmbeddedLanguageInfo _info;
+        private readonly SemanticModel _semanticModel;
+        private readonly INamedTypeSymbol _dateTimeType;
+        private readonly INamedTypeSymbol _dateTimeOffsetType;
+
+        public DateAndTimePatternDetector(
+            SemanticModel semanticModel,
+            EmbeddedLanguageInfo info,
+            INamedTypeSymbol dateTimeType,
+            INamedTypeSymbol dateTimeOffsetType)
+        {
+            _info = info;
+            _semanticModel = semanticModel;
+            _dateTimeType = dateTimeType;
+            _dateTimeOffsetType = dateTimeOffsetType;
+        }
+
+        public static DateAndTimePatternDetector? TryGetOrCreate(
+            SemanticModel semanticModel, EmbeddedLanguageInfo info)
+        {
+            // Do a quick non-allocating check first.
+            if (_modelToDetector.TryGetValue(semanticModel, out var detector))
+            {
+                return detector;
+            }
+
+            return _modelToDetector.GetValue(
+                semanticModel, _ => TryCreate(semanticModel, info));
+        }
+
+        private static DateAndTimePatternDetector? TryCreate(
+            SemanticModel semanticModel, EmbeddedLanguageInfo info)
+        {
+            var dateTimeType = semanticModel.Compilation.GetTypeByMetadataName(typeof(System.DateTime).FullName!);
+            var dateTimeOffsetType = semanticModel.Compilation.GetTypeByMetadataName(typeof(System.DateTimeOffset).FullName!);
+            if (dateTimeType == null || dateTimeOffsetType == null)
+                return null;
+
+            return new DateAndTimePatternDetector(semanticModel, info, dateTimeType, dateTimeOffsetType);
+        }
+
+        public static bool IsPossiblyDateAndTimeToken(
```

Comment:
My previous comment on this no longer applies- but could there be a comment on what this is passing into the out params?  e.g. argumentNode is the argument where the date time string literal is passed, similar for invocationExpression

Reply:
sure.

Reply:
doc'ed